### DISCUSSION TITLE
Add a reference to a node in a different server's address space

### DISCUSSION THREAD

**Discussion by FMeinicke** [FreeOpcUa/opcua-asyncio#1697] [Q&A]
Hi,

I would like to be able to add a reference to a node that is not in the address space of the local server but in the address space of another server that is available in the same network.

According to the OPC UA specification, this should be possible by adding the URI of the other server to the *ServerArray* property of the local server (https://reference.opcfoundation.org/Core/Part5/v105/docs/6.3.1). Then I should be able to use an *ExpandedNodeId* with the *ServerIndex* set to the index of that server in the *ServerArray* to identify any node in the address space of the other server.

Looking at the code, I think this is currently not yet possible. I also found FreeOpcUa/python-opcua#557 which mentions the same thing but for the sync-only version of the library. That issue wasn't resolved, so I believe there hasn't been done any work on this feature, yet.

Since I don't know, how this feature could be implemented, I'd like to ask if there is someone who would be willing to invest time in adding this feature.

Secondly, provided that this feature is implemented, what would I have to use as the URN of the other server in the *ServerArray*? Is that the URL of the other server (i.e. `myotherserver.hostname:4840`, maybe even with the `opc.tcp://` protocol), or just its domain name (`myotherserver.hostname`)?  
I guess the answer to this question depends on the implementation, so this can probably only be answered once the feature is implemented...

Thanks in advance!

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/Core/Part3/v105/docs/4.4.4

> The [TargetNode](https://reference.opcfoundation.org/search/356?t=TargetNode) of a [Reference](https://reference.opcfoundation.org/search/356?t=Reference) may be in the same [AddressSpace](https://reference.opcfoundation.org/search/356?t=AddressSpace) or in the [AddressSpace](https://reference.opcfoundation.org/search/356?t=AddressSpace) of another OPC UA [Server](https://reference.opcfoundation.org/search/356?t=Server). [TargetNodes](https://reference.opcfoundation.org/search/356?t=TargetNodes) located in other [Servers](https://reference.opcfoundation.org/search/356?t=Servers) are identified in OPC UA [Services](https://reference.opcfoundation.org/search/356?t=Services) using a combination of the remote [Server](https://reference.opcfoundation.org/search/356?t=Server) name and the identifier assigned to the [Node](https://reference.opcfoundation.org/search/356?t=Node) by the remote [Server](https://reference.opcfoundation.org/search/356?t=Server).

> OPC UA does not require that [the](https://reference.opcfoundation.org/search/356?t=the) [TargetNode ](https://reference.opcfoundation.org/search/356?t=TargetNode)exists, thus [References](https://reference.opcfoundation.org/search/356?t=References) may point to a [Node](https://reference.opcfoundation.org/search/356?t=Node) that does not exist.

So you can just add the server to the ServerArray ("urn:<your own urn here!>") and use the index of the added server in the expanded nodeid!

Index 0 is your own "urn:" (AppUri) and not the remote so you should have at least index 0 and index 1 after adding!

https://reference.opcfoundation.org/Core/Part4/v104/docs/7.11

> Index that identifies the [Server](https://reference.opcfoundation.org/search/17?t=Server) that contains the [TargetNode](https://reference.opcfoundation.org/search/17?t=TargetNode). This [Server](https://reference.opcfoundation.org/search/17?t=Server) may be the local [Server](https://reference.opcfoundation.org/search/17?t=Server) or a remote [Server](https://reference.opcfoundation.org/search/17?t=Server).

> This index is the index of that [Server](https://reference.opcfoundation.org/search/17?t=Server) in the local [Server](https://reference.opcfoundation.org/search/17?t=Server)’s [Server](https://reference.opcfoundation.org/search/17?t=Server) table. The index of the local [Server](https://reference.opcfoundation.org/search/17?t=Server) in the [Server](https://reference.opcfoundation.org/search/17?t=Server) table is always 0. All remote [Servers](https://reference.opcfoundation.org/search/17?t=Servers) have indexes greater than 0. The [Server](https://reference.opcfoundation.org/search/17?t=Server) table is contained in the [Server](https://reference.opcfoundation.org/search/17?t=Server) [Object](https://reference.opcfoundation.org/search/17?t=Object) in the [AddressSpace](https://reference.opcfoundation.org/search/17?t=AddressSpace) (see [OPC 10000-3](https://reference.opcfoundation.org/Core/Part4/v104/docs/?r=UAPart3) and [OPC 10000-5](https://reference.opcfoundation.org/Core/Part4/v104/docs/?r=UAPart5)).

> The [Client](https://reference.opcfoundation.org/search/17?t=Client) may read the [Server](https://reference.opcfoundation.org/search/17?t=Server) table [Variable](https://reference.opcfoundation.org/search/17?t=Variable) to access the description of the target [Server.](https://reference.opcfoundation.org/search/17?t=Server.)

**Reply by AndreasHeine** [reply]
the only thing whats missing is a convinience function for doing that server index adding, if i look at src:
https://github.com/FreeOpcUa/opcua-asyncio/blob/1fe61dfe279b5d9722d11bb6ec191dfa17ad2727/asyncua/common/node.py#L793

**Reply by FMeinicke** [reply]
> So you can just add the server to the ServerArray ("urn:<your own urn here!>") and use the index of the added server in the expanded nodeid!
> 
> Index 0 is your own "urn:" (AppUri) and not the remote so you should have at least index 0 and index 1 after adding!

Right, thanks for confirming!

My idea is then to do something like this:

```python
server = Server()
app_uri = "urn:my:opcua.server"
other_app_uri = "urn:my:opcua.other_server"
await server.set_application_uri(app_uri)

async with server:
    server_array_node = Node(server.iserver.isession, ua.ObjectIds.Server_ServerArray)
    await server_array.write_value([app_uri, other_app_uri])

    node_in_other_server = ua.ExpandedNodeId("OtherServerNode", 2, ServerIndex=1)
    await server.nodes.objects.add_reference(node_in_other_server , ua.ObjectIds.HasComponent)
```

This, of course, currently gives a `BadTargetNodeIdInvalid` error because `add_reference()` cannot find the NodeId in the local server address space and currently it doesn't look into other server's address spaces:

https://github.com/FreeOpcUa/opcua-asyncio/blob/1fe61dfe279b5d9722d11bb6ec191dfa17ad2727/asyncua/server/address_space.py#L441-L449

Do you think it would be difficult to implement this? I would try to play around myself but I don't know how the server would "resolve" the URN of the other server to something (the endpoint URI?) it can use to access the address space.

**Reply by AndreasHeine** [reply]
probably read the serverarray first, check if its alread in the array if not append and write ;)

looking into others server addresspsace is not described in OPC UA Spec. or did i miss that part!?
> OPC UA does not require that [the](https://reference.opcfoundation.org/search/356?t=the) [TargetNode ](https://reference.opcfoundation.org/search/356?t=TargetNode)exists, thus [References](https://reference.opcfoundation.org/search/356?t=References) may point to a [Node](https://reference.opcfoundation.org/search/356?t=Node) that does not exist.

> The [Client](https://reference.opcfoundation.org/search/17?t=Client) may read the [Server](https://reference.opcfoundation.org/search/17?t=Server) table [Variable](https://reference.opcfoundation.org/search/17?t=Variable) to access the description of the target [Server.](https://reference.opcfoundation.org/search/17?t=Server.)

"THE CLIENT MAY READ THE..." but that requires a complete new client instance (with user and so on besides networking) and a lot of logic!?

**Reply by AndreasHeine** [reply]
@FMeinicke 
```python
if addref.TargetNodeId not in self._aspace: 
         return ua.StatusCode(ua.StatusCodes.BadTargetNodeIdInvalid) 
```
probably this is a false check according to spec.!


```python
if addref.TargetNodeId not in self._aspace: 
         # only return Error if ServerIndex is 0 !!!
         return ua.StatusCode(ua.StatusCodes.BadTargetNodeIdInvalid) 
```

**Reply by FMeinicke** [reply]
If I understand correctly, then `add_reference()` would need to accept NodeIds that point to other server's address spaces.  
Then a client (e.g. UaExpert) when it browses the address space of server1 sees that reference to the node in server2's address space and it (the client, so UaExpert) must figure out how to "resolve" the URN of server. Is that what the spec says?  
(To be honest, I'm not that deep into the whole OPC UA spec. I saw that there is this feature where a server can reference nodes in other servers, which is something that I would need for the device I'm currently modeling in OPC UA.)

**Reply by FMeinicke** [reply]
> ```python
> if addref.TargetNodeId not in self._aspace: 
>          return ua.StatusCode(ua.StatusCodes.BadTargetNodeIdInvalid) 
> ```
> 
> probably this is a false check according to spec.!
> 
> ```python
> if addref.TargetNodeId not in self._aspace: 
>          # only return Error if ServerIndex is 0 !!!
>          return ua.StatusCode(ua.StatusCodes.BadTargetNodeIdInvalid) 
> ```

Yes, that's what I meant with *`add_reference()` would need to accept NodeIds that point to other server's address spaces.*

**Reply by AndreasHeine** [reply]
> If I understand correctly, then `add_reference()` would need to accept NodeIds that point to other server's address spaces. Then a client (e.g. UaExpert) when it browses the address space of server1 sees that reference to the node in server2's address space and it (the client, so UaExpert) must figure out how to "resolve" the URN of server. Is that what the spec says? (To be honest, I'm not that deep into the whole OPC UA spec. I saw that there is this feature where a server can reference nodes in other servers, which is something that I would need for the device I'm currently modeling in OPC UA.)

thats what i know you just get informed with the ServerIndex that its not in this server and usually via opc ua global/local dicovery service you have to figure out with that "urn" is registered at the discovery server and maybe what the actual endpoint is -> from that you need to create a new client to do whatever you want to do (dont think spec defined the logic for this the number of usecases is infinite)

**Reply by FMeinicke** [reply]
Ok, I see. So I probably need a discovery server in the network alongside the two servers that reference each other, so that clients can ask the discovery server how they can connect to the server with `"urn:my:opcua.other_server"`.

**Reply by AndreasHeine** [reply]
@FMeinicke yes
